### PR TITLE
Changes the Reddit Pixel to support Oauth2

### PR DESCRIPTION
**Pull request by renatost**
Apparently Reddit no longer supports API Key to access the conversion API. This change implements the oauth2 authentication mechanism as described in the [documentation]( https://ads-api.reddit.com/docs/v2/#section/Authentication/Reddit%20API%20Key)

### CODE REVIEW

File: template.tpl [hunk 1]
```diff
@@ -391,39 +406,123 @@ if (rdtcid) {
...
 
-sendHttpRequest(
-  postUrl,
-  (statusCode, headers, body) => {
+
+
+
+Firestore.read('reddit/auth_key').then((result) => {
+  if (result.reason == "not_found") {
+    refreshKey().then(r => sendHttpRequest(r));
```
Comment (Bukashk0zzz) [inline template.tpl:414]: sendHttpRequest you probably mean sendRequest function here

File: template.tpl [hunk 2]
```diff
@@ -391,39 +406,123 @@ if (rdtcid) {
...
+  }
+  const authKey = result.data;
+  if (authKey.lastUpdated < (getTimestampMillis() - 1000 * 60 * 60)) {
+    refreshKey().then(r => sendHttpRequest(r));
```
Comment (Bukashk0zzz) [inline template.tpl:418]: sendHttpRequest you probably mean sendRequest function here

File: template.tpl [hunk 3]
```diff
@@ -391,39 +406,123 @@ if (rdtcid) {
...
+  } else {
+    sendRequest(authKey.apiKey);
+  }  
+}); 
+
+
+
+function sendRequest(authKey) {  
+  if (isLoggingEnabled) {
+    logToConsole(
```
Comment (Bukashk0zzz) [inline template.tpl:428]: please use logs standard https://github.com/stape-io/gtm-standards/blob/main/standards/GTMS-3.md

name AuthKey not allowed here

File: template.tpl [hunk 4]
```diff
@@ -391,39 +406,123 @@ if (rdtcid) {
...
         JSON.stringify({
-          Name: 'Reddit',
-          Type: 'Response',
-          TraceId: traceId,
-          EventName: eventType.tracking_type === 'Custom' ? eventType.custom_event_name : eventType.tracking_type,
-          ResponseStatusCode: statusCode,
-          ResponseHeaders: headers,
-          ResponseBody: body,
+        Name: 'Reddit',
+        Type: 'RequestRefreshKey',
```
Comment (Bukashk0zzz) [inline template.tpl:482]: please use logs standard https://github.com/stape-io/gtm-standards/blob/main/standards/GTMS-3.md

Type here must be Request

File: template.tpl [hunk 5]
```diff
@@ -391,39 +406,123 @@ if (rdtcid) {
...
+        TraceId: traceId,
+        EventName: eventType.tracking_type === 'Custom' ? eventType.custom_event_name : eventType.tracking_type,
```
Comment (Bukashk0zzz) [inline template.tpl:484]: Name RefreshKey

File: template.tpl [hunk 6]
```diff
@@ -391,39 +406,123 @@ if (rdtcid) {
...
+          'Content-Type': 'application/x-www-form-urlencoded'
+        }
+      }, "grant_type=refresh_token&refresh_token=" + data.refreshToken);
+      
+      httpPromise.then( result => {
+        if (isLoggingEnabled) {
+          logToConsole(
+            JSON.stringify({
+              Name: 'Reddit',
+              Type: 'ResponseRefreshKey',
```
Comment (Bukashk0zzz) [inline template.tpl:508]: please use logs standard https://github.com/stape-io/gtm-standards/blob/main/standards/GTMS-3.md

Type here must be Response

File: template.tpl [hunk 7]
```diff
@@ -391,39 +406,123 @@ if (rdtcid) {
...
+              TraceId: traceId,
+              EventName: eventType.tracking_type === 'Custom' ? eventType.custom_event_name : eventType.tracking_type,
```
Comment (Bukashk0zzz) [inline template.tpl:510]: Name RefreshKey

File: template.tpl [hunk 8]
```diff
@@ -391,39 +406,123 @@ if (rdtcid) {
...
+              ResponseStatusCode: result.statusCode,
+              ResponseHeaders: result.headers,
+              ResponseBody: result.body,
+            })
+          );
+        }        
+        const body = JSON.parse(result.body).access_token;
+        Firestore.write('reddit/auth_key', {apiKey: body, lastUpdated: getTimestampMillis()});
```
Comment (Bukashk0zzz) [inline template.tpl:518]: Please use this standard for Firestor setting https://github.com/stape-io/gtm-standards/blob/main/standards/GTMS-6.md

File: template.tpl [hunk 9]
```diff
@@ -391,39 +406,123 @@ if (rdtcid) {
...
+        'Authorization': "Bearer " + authKey 
+      },
+      method: 'POST',
+    },
+    JSON.stringify(postBody)
+    );
+}
+
+function refreshKey() {
+  return Promise.create((res, rej) => {
```
Comment (Bukashk0zzz) [inline template.tpl:477]: that's not required to use Promise here as sendHttpRequest automatically create it for you.
but you can leave it as is.

**Review by Bukashk0zzz — CHANGES_REQUESTED** [review]
Thanks for a good work and sharing.
Please resolve issues in comments.

File: template.tpl (hunk 3 above)
Comment (renatost) [reply template.tpl:457]: Removed AuthKey

File: template.tpl (hunk 9 above)
Comment (renatost) [reply template.tpl:505]: Using that created promise to write to firebase before returning. After I created it, realized I don't need to wait for firebase to complete.

File: template.tpl (hunk 5 above)
Comment (renatost) [reply template.tpl:512]: I am assuming you mean Name: 'Reddit' and type 'Request' as outlined in the standards doc, correct?

File: template.tpl (hunk 7 above)
Comment (renatost) [reply template.tpl:538]: I am assuming you mean Name: 'Reddit' and type 'Response' as outlined in the standards doc, correct?

**Review by renatost — COMMENTED** [review]
glad to be able to collaborate.

File: template.tpl (hunk 5 above)
Comment (Bukashk0zzz) [reply template.tpl:512]: EventName should be RefreshKey

File: template.tpl (hunk 7 above)
Comment (Bukashk0zzz) [reply template.tpl:538]: EventName should be RefreshKey

File: template.tpl [hunk 10]
```diff
@@ -389,41 +433,124 @@ if (rdtcid) {
...
+
+Firestore.read(data.firebasePath, firebaseOptions).then((result) => {
+  if (result.reason == "not_found") {
+    refreshKey().then(r => sendRequest(r));
+  }
+  const authKey = result.data;
+  if (authKey.lastUpdated < (getTimestampMillis() - 1000 * 60 * 60)) {
+    refreshKey().then(r => sendRequest(r));
+  } else {
+    sendRequest(authKey.apiKey);
```
Comment (Bukashk0zzz) [inline template.tpl:449]: if result.reason == "not_found" then 
sendRequest will be used 2 times and conversion will be ended 2 times

File: template.tpl [hunk 11]
```diff
@@ -389,41 +433,124 @@ if (rdtcid) {
...
     httpOnly: false,
   });
+}  
+
+
+
+
+Firestore.read(data.firebasePath, firebaseOptions).then((result) => {
+  if (result.reason == "not_found") {
+    refreshKey().then(r => sendRequest(r));
```
Comment (Bukashk0zzz) [inline template.tpl:443]: I think here should be return with data.gtmOnSuccess();

File: template.tpl [hunk 12]
```diff
@@ -389,41 +433,124 @@ if (rdtcid) {
...
+  }
+  const authKey = result.data;
+  if (authKey.lastUpdated < (getTimestampMillis() - 1000 * 60 * 60)) {
+    refreshKey().then(r => sendRequest(r));
+  } else {
+    sendRequest(authKey.apiKey);
+  }  
+}, () => refreshKey().then(r => sendRequest(r)));
```
Comment (Bukashk0zzz) [inline template.tpl:451]: I think here should be data.gtmOnFailure

**Review by Bukashk0zzz — CHANGES_REQUESTED** [review]

File: template.tpl (hunk 10 above)
Comment (renatost) [reply template.tpl:450]: Added a return here.

File: template.tpl (hunk 11 above)
Comment (renatost) [reply template.tpl:443]: data.gtmOnSuccess() is called inside sendRequest -- do we need to call it here again?

File: template.tpl (hunk 12 above)
Comment (renatost) [reply template.tpl:452]: data.gtmOnFailure() is called inside sendRequest -- do we need to call it here again?

File: template.tpl (hunk 10 above)
Comment (renatost) [reply template.tpl:450]: added return here

File: template.tpl (hunk 11 above)
Comment (renatost) [reply template.tpl:443]: data.gtmOnSuccess() is called inside sendRequest -- do we need to call it here again?

File: template.tpl (hunk 12 above)
Comment (renatost) [reply template.tpl:452]: data.gtmOnFailure() is called inside sendRequest -- do we need to call it here again?

**Review by Bukashk0zzz — APPROVED** [review]

**Pull request merged**